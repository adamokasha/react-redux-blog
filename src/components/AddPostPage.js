import React from "react";
import { connect } from "react-redux";

import PostForm from "./PostForm";
import { startAddPost } from "../actions/posts";

export class AddPostPage extends React.Component {
  constructor(props) {
    super(props);

    this.state = {
      error: ""
    };
  }
  componentDidMount() {
    window.scrollTo(0, 0);
  }
  renderError = () => {
    if (this.state.error) {
      return (
        <div className="error-msg alert alert-danger" role="alert">
          <p>{this.state.error}</p>
        </div>
      );
    }
  };
  onSubmit = async post => {
    try {
      await this.props.startAddPost(post);
      this.props.history.push("/");
    } catch (e) {
      this.setState({ error: e.message });
    }
  };
  render() {
    return (
      <div className="container">
        <h3 className="text-center mt-4">Add Post</h3>
        {this.renderError()}
        <PostForm onSubmit={this.onSubmit} />
      </div>
    );
  }
}

const mapDispatchToProps = dispatch => ({
  startAddPost: post => dispatch(startAddPost(post))
});

export default connect(
  null,
  mapDispatchToProps
)(AddPostPage);
