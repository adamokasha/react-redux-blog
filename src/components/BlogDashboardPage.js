import React from "react";
import { connect } from "react-redux";

import Header from "./Header";
import PostsList from "./PostsList";

export const BlogDashboardPage = ({ displayName }) => {
  function renderWelcome() {
    if (displayName) {
      return <h3>Welcome {displayName}.</h3>;
    }
    return;
  }
  return (
    <div>
      <Header />
      <div id="landing">
        <div className="primary-overlay">
          <div className="container h-100">
            <div className="row hero-text">
              <div className="col-lg-7">
                <h1 className="text-white mb-5">
                  React Blog - Responsive blog styled with Bootstrap 4
                </h1>
                <a
                  href="https://github.com/adamokasha/react-redux-blog"
                  target="_blank"
                  className="text-uppercase text-white hero-link"
                >
                  <u>Learn More</u>
                </a>
              </div>
            </div>
          </div>
        </div>
      </div>
      <br />
      <div className="container mb-3">{renderWelcome()}</div>
      <div>
        <PostsList />
      </div>
    </div>
  );
};

const mapStateToProps = state => ({
  displayName: state.auth.displayName
});

export default connect(mapStateToProps)(BlogDashboardPage);
