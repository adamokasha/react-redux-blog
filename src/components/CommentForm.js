import React from "react";
import { connect } from "react-redux";

import { startAddComment } from "../actions/posts";

export class CommentForm extends React.Component {
  constructor(props) {
    super(props);

    this.state = {
      comment: "",
      error: ""
    };
  }
  onCommentChange = e => {
    this.setState({ comment: e.target.value });
  };
  renderError = () => {
    if (this.state.error) {
      return <div className="error-msg text-danger">{this.state.error}</div>;
    }
  };
  onSubmit = e => {
    e.preventDefault();

    const { comment } = this.state;
    if (comment.length < 8 || comment.length > 128) {
      this.setState({ error: "Comment must be between 8 and 128 characters." });
    } else {
      this.setState({ error: "" });
      this.props.startAddComment(this.props.id, this.state.comment);
      this.setState({ comment: "" });
    }
  };
  render() {
    return (
      <div className="mb-3">
        {this.renderError()}
        <form onSubmit={this.onSubmit}>
          <div className="form-group col-md-4 pl-0 mb-2">
            <textarea
              className="form-control"
              id="comment-textarea"
              placeholder="Leave a comment..."
              onChange={this.onCommentChange}
              maxLength={128}
              rows="3"
              value={this.state.comment}
            />
          </div>
          <button className="btn btn-primary">Add Comment</button>
        </form>
      </div>
    );
  }
}

const mapDispatchToProps = dispatch => ({
  startAddComment: (id, comment) => dispatch(startAddComment(id, comment))
});

export default connect(
  null,
  mapDispatchToProps
)(CommentForm);
