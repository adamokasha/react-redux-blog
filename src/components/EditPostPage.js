import React from "react";
import { connect } from "react-redux";

import PostForm from "./PostForm";
import { startEditPost } from "../actions/posts";
import { startDeletePost } from "../actions/posts";

export class EditPostPage extends React.Component {
  constructor(props) {
    super(props);

    this.state = {
      error: ""
    };
  }
  componentDidMount() {
    window.scrollTo(0, 0);
  }
  renderError = () => {
    if (this.state.error) {
      return (
        <div className="error-msg">
          <p>{this.state.error}</p>
        </div>
      );
    }
  };
  onSubmit = async post => {
    try {
      await this.props.startEditPost(post, this.props.post._id);
      this.props.history.push("/");
    } catch (e) {
      this.setState({ error: e.message });
    }
  };
  deletePost = async () => {
    try {
      await this.props.startDeletePost(this.props.post._id);
      this.props.history.push("/");
    } catch (e) {
      this.setState({ error: e.message });
    }
  };
  render() {
    return (
      <div className="container">
        <h3 className="text-center mt-4">Edit Post</h3>
        <button
          className="btn-delete btn btn-dark btn-sm float-right"
          onClick={this.deletePost}
        >
          <i className="far fa-trash-alt mr-2" />
          Delete Post
        </button>
        {this.renderError()}
        <PostForm post={this.props.post} onSubmit={this.onSubmit} />
      </div>
    );
  }
}

const mapStateToProps = (state, props) => ({
  post: state.posts.find(post => post._id === props.match.params.id)
});

const mapDispatchToProps = dispatch => ({
  startEditPost: (post, id) => dispatch(startEditPost(post, id)),
  startDeletePost: id => dispatch(startDeletePost(id))
});

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(EditPostPage);
