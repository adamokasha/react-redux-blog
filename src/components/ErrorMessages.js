import React from 'react';

class ErrorMessages extends React.Component {
  constructor(props) {
    super(props);
  }
  componentDidMount() {
    window.scrollTo(0, 0);
  }
  render() {
    return (
      <ul className="error-list list-group mb-3">
        {this.props.errors.map(error => {
          return (
            <li className="list-group-item list-group-item-danger">{error}</li>
          );
        })}
      </ul>
    );
  }
}

export default ErrorMessages;
