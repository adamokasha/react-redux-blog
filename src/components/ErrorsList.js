import React from 'react';

class ErrorsList extends React.Component {
  constructor(props) {
    super(props);
  }
  componentDidMount() {
    window.scrollTo(0, 0);
  }
  render() {
    return (
      <ul className="error-list list-group mb-3">
        {this.props.errors.map((error, i) => {
          return (
            <li key={i} className="list-group-item list-group-item-danger">{error}</li>
          );
        })}
      </ul>
    );
  }
}

export default ErrorsList;
