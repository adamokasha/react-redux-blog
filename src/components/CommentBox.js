import React from "react";
import { connect } from "react-redux";
import moment from "moment";

import CommentForm from "./CommentForm";
import { sortCommentsByDate } from "../selectors/posts";

export class CommentBox extends React.Component {
  renderCommentForm = () => {
    if (this.props.auth.id) {
      return <CommentForm id={this.props.id} />;
    }
    return <p className="font-italic">You must be signed in to comment...</p>;
  };
  renderComments = () => {
    if (this.props.comments.length > 0) {
      return sortCommentsByDate(this.props.comments).map(comment => {
        return (
          <li className="list-group-item pl-0" key={comment._id}>
            <p className="text-muted mb-1">
              Posted by {comment.createdBy} on{" "}
              {moment(comment.date).format("MM-DD-YY [at] HH:mm")}
            </p>
            <p className="mb-1">{comment.comment}</p>
          </li>
        );
      });
    } else {
      return <p>Be the first to comment...</p>;
    }
  };
  render() {
    return (
      <div>
        <h4>Comments</h4>
        {this.renderCommentForm()}
        <ul className="list-group-flush pl-0">{this.renderComments()}</ul>
      </div>
    );
  }
}

const mapStateToProps = state => ({
  auth: state.auth
});

export default connect(mapStateToProps)(CommentBox);
