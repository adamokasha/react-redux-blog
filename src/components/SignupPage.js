import React from 'react';
import { connect } from 'react-redux';
import validator from 'validator';

import ErrorsList from './ErrorsList';
import { startSignup } from '../actions/auth';

export class SignupPage extends React.Component {
  constructor(props) {
    super(props);

    this.state = {
      displayName: '',
      email: '',
      password: '',
      passwordConfirm: '',
      errors: null
    };
  }
  componentDidMount() {
    window.scrollTo(0, 0);
  }
  onDisplayNameChange = e => {
    this.setState({ displayName: e.target.value });
  };
  onEmailChange = e => {
    this.setState({ email: e.target.value });
  };
  onPasswordChange = e => {
    this.setState({ password: e.target.value });
  };
  onPasswordConfirmChange = e => {
    this.setState({ passwordConfirm: e.target.value });
  };
  renderErrors = () => {
    const { errors } = this.state;
    if (errors && errors.length >= 1) {
      return <ErrorsList errors={errors} />;
    }
  };
  onSubmit = async e => {
    e.preventDefault();

    const { displayName, email, password, passwordConfirm } = this.state;
    const errors = [];

    if (displayName.length < 6 || displayName.length > 12) {
      errors.push('Display name must be between 6 and 12 characters.');
    }
    if (!validator.isEmail(email)) {
      errors.push('Please enter a valid email.');
    }
    if (password.length < 6 || password.length > 12) {
      errors.push('Password must be between 6 and 12 characters');
    }
    if (!validator.equals(password, passwordConfirm)) {
      errors.push('Passwords must match');
    }

    if (errors.length >= 1) {
      this.setState({ errors });
    } else {
      try {
        await this.props.startSignup(displayName, email, password);
        this.props.history.push('/');
      } catch (e) {
        const errors = [];
        Object.keys(e).forEach(key => {
          errors.push(e[key].message);
        });
        this.setState({ errors });
      }
    }
  };
  render() {
    return (
      <div className="container mt-4">
        <div className="row">
          <div className="container col-lg-5">
            <h3 className="mb-3 text-center">Sign Up</h3>
            {this.renderErrors()}
            <form onSubmit={this.onSubmit}>
              <div className="form-group">
                <label htmlFor="displayname">Display Name:</label>
                <input
                  className="form-control"
                  type="text"
                  name="displayname"
                  placeholder="Enter a 6-12 character display name"
                  onChange={this.onDisplayNameChange}
                  autoFocus
                />
              </div>
              <div className="form-group">
                <label htmlFor="email">Email:</label>
                <input
                  className="form-control"
                  type="text"
                  name="email"
                  placeholder="Enter a valid email"
                  onChange={this.onEmailChange}
                />
              </div>
              <div className="form-group">
                <label htmlFor="password">Password:</label>
                <input
                  className="form-control"
                  type="password"
                  name="password"
                  placeholder="Enter a 6-12 character password"
                  onChange={this.onPasswordChange}
                />
              </div>
              <div className="form-group">
                <label htmlFor="passwordConfirm">Confirm Password:</label>
                <input
                  className="form-control"
                  type="password"
                  name="passwordConfirm"
                  placeholder="Re-enter your password"
                  onChange={this.onPasswordConfirmChange}
                />
              </div>
              <div className="form-group">
                <button className="btn btn-primary btn-block">Signup</button>
              </div>
            </form>
          </div>
        </div>
      </div>
    );
  }
}

const mapDispatchToProps = dispatch => ({
  startSignup: (displayName, email, password) =>
    dispatch(startSignup(displayName, email, password))
});

export default connect(
  null,
  mapDispatchToProps
)(SignupPage);
