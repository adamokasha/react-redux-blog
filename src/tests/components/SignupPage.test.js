import React from "react";
import { shallow, configure } from "enzyme";
import { mount } from "enzyme";
import Adapter from "enzyme-adapter-react-16";

import { SignupPage } from "../../components/SignupPage";

configure({ adapter: new Adapter() });

test("should render SignupPage correctly", () => {
  const wrapper = shallow(<SignupPage />);
  expect(wrapper).toMatchSnapshot();
});

test("should render error if form input invalid", () => {
  const wrapper = mount(<SignupPage />);
  wrapper.find("form").simulate("submit", {
    preventDefault: () => {}
  });
  expect(wrapper.state("errors")).toBeTruthy();
  expect(wrapper.find(".list-group-item-danger").length).toBeGreaterThan(0);
});

test("should set displayName on input change", () => {
  const value = "aNewUser";
  const wrapper = shallow(<SignupPage />);
  wrapper.find('input[name="displayname"]').simulate("change", {
    target: { value }
  });
  expect(wrapper.state("displayName")).toBe(value);
});

test("should set email on input change", () => {
  const value = "email@example.com";
  const wrapper = shallow(<SignupPage />);
  wrapper.find('input[name="email"]').simulate("change", {
    target: { value }
  });
  expect(wrapper.state("email")).toBe(value);
});

test("should set password on input change", () => {
  const value = "123abc";
  const wrapper = shallow(<SignupPage />);
  wrapper.find('input[name="password"]').simulate("change", {
    target: { value }
  });
  expect(wrapper.state("password")).toBe(value);
});

test("should call startSignup if form input completely valid", () => {
  const user = {
    displayName: "validUser",
    email: "valid@email.com",
    password: "123abc",
    passwordConfirm: "123abc"
  };
  const startSignup = jest.fn();
  const wrapper = shallow(<SignupPage startSignup={startSignup} />);
  wrapper.setState(user);
  wrapper.find("form").simulate("submit", {
    preventDefault: () => {}
  });
  const { displayName, email, password } = user;
  expect(startSignup).toHaveBeenLastCalledWith(displayName, email, password);
});
