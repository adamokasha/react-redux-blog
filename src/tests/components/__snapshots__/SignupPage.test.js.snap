// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render SignupPage correctly 1`] = `
<div
  className="container mt-4"
>
  <div
    className="row"
  >
    <div
      className="container col-lg-5"
    >
      <h3
        className="mb-3 text-center"
      >
        Sign Up
      </h3>
      <form
        onSubmit={[Function]}
      >
        <div
          className="form-group"
        >
          <label
            htmlFor="displayname"
          >
            Display Name:
          </label>
          <input
            autoFocus={true}
            className="form-control"
            name="displayname"
            onChange={[Function]}
            placeholder="Enter a 6-12 character display name"
            type="text"
          />
        </div>
        <div
          className="form-group"
        >
          <label
            htmlFor="email"
          >
            Email:
          </label>
          <input
            className="form-control"
            name="email"
            onChange={[Function]}
            placeholder="Enter a valid email"
            type="text"
          />
        </div>
        <div
          className="form-group"
        >
          <label
            htmlFor="password"
          >
            Password:
          </label>
          <input
            className="form-control"
            name="password"
            onChange={[Function]}
            placeholder="Enter a 6-12 character password"
            type="password"
          />
        </div>
        <div
          className="form-group"
        >
          <label
            htmlFor="passwordConfirm"
          >
            Confirm Password:
          </label>
          <input
            className="form-control"
            name="passwordConfirm"
            onChange={[Function]}
            placeholder="Re-enter your password"
            type="password"
          />
        </div>
        <div
          className="form-group"
        >
          <button
            className="btn btn-primary btn-block"
          >
            Signup
          </button>
        </div>
      </form>
    </div>
  </div>
</div>
`;
